### PR TITLE
Optimizers support for parameter groups

### PR DESCRIPTION
**Pull request by NiklasGustafsson**
This is a massive PR.

The main thrust is to support parameter groups in the optimizers and LR schedulers. This is a massive and breaking change.

The change also involves moving all the optimizer implementations to managed code, except for LBFGS, which does not allow more than a single parameter group, and therefore does not require the move.

On the way to implementing this support, a few things were discovered missing and/or incorrect, and were therefore fixed.

@lostmsu, @oluwandabira -- if you have a chance to take a look, too, that would be wonderful. I still haven't heard about the right way to include individuals in the 'Request Review' UI.

### CODE REVIEW

**Comment by NiklasGustafsson** [comment]
@lostmsu:

Yes, excellent question. 

I completely agree that the fact that we're just interfacing with native code is attractive. Without a question, it is why we have come as far as we have as quickly as we have. There are a couple of areas, like determining whether types are float, complex, etc. doesn't seem worth going to C++ for, but otherwise, TorchSharp is and should remain, a .NET wrapper of LibTorch.

There are only (even before this PR) three areas where there is significant managed code:

1. TorchVision (but I moved code to C++ to avoid the dispose problem)
2. Optimizers (half of them were already in managed code)
3. LR Schedulers.

The situation with the optimizers is that only about half of the ones available in 1.10 are available in native code, so there was already a number of managed code implementations. In fact, I dragged my feet for a long time, some months back, because I didn't want to tackle the managed code optimizers -- this was before the DisposeScope addition, when expressing tensor computations in .NET was nasty.

So, faced with a disruptive change, i.e. adding parameter groups, and the challenge of representing / encoding parameter groups and their options in two different ways, and having to continue to maintain two very different approaches to implementation, I felt that it would be better to have just one, like Pytorch does.

So, that's the rationale, for better or worse.

File: src/TorchSharp/NN/Module.cs [hunk 1]
```diff
@@ -151,7 +153,9 @@ public virtual Module to(DeviceType deviceType, int deviceIndex = -1)
 
                             if (param is not null) {  // This test must come before the Tensor test
                                 if (deviceType != param.device_type || deviceIndex != param.device_index) {
-                                    var p = new Modules.Parameter(param.to(deviceType, deviceIndex), param.requires_grad);
+                                    var t = param.to(deviceType, deviceIndex);
+                                    t.retain_grad();
```
Comment (lostmsu) [inline src/TorchSharp/NN/Module.cs:157]: Can you comment on the need for `retain_grad` call here? Is there a bug this fixes?

File: src/TorchSharp/NN/LRScheduler.cs [hunk 2]
```diff
@@ -26,16 +26,18 @@ protected LRScheduler()
 
                     }
 
-                    protected LRScheduler(ILearningRateController optimizer, int last_epoch = -1, bool verbose = false)
+                    protected LRScheduler(Optimizer optimizer, int last_epoch = -1, bool verbose = false)
```
Comment (lostmsu) [inline src/TorchSharp/NN/LRScheduler.cs:29]: It would seem more natural to have `LRScheduler` work per parameter group, not per optimizer. Same for all other schedulers.

File: src/TorchSharp/NN/Optimizer.cs [hunk 3]
```diff
@@ -293,563 +452,1901 @@ public static AdadeltaOptimizer Adadelta(IEnumerable<(string name, Modules.Param
...
+                        foreach (var group in _parameter_groups) {
 
-                    using var grad = p.grad();
+                            body(group as T);
 
-                    if (grad is null) continue;
+                        }
 
-                    grad.zero_().Dispose();
+                        d.DisposeEverything();
```
Comment (lostmsu) [inline src/TorchSharp/NN/Optimizer.cs]: Why do you have to call this manually? I thought you did not need to, it would be called by the implicit `d.Dispose()`?

File: src/TorchSharp/NN/Optimizer.cs [hunk 4]
```diff
@@ -293,563 +452,1901 @@ public static AdadeltaOptimizer Adadelta(IEnumerable<(string name, Modules.Param
...
+            {
+                base.Parameters = parameters;
+                base.Options = options;
+            }
+
+            public new TOptions Options { get => (TOptions)base.Options; set => base.Options = value; }
+
+        }
+
+        public class SGD : OptimizerHelper, IMomentum
```
Comment (lostmsu) [inline src/TorchSharp/NN/Optimizer.cs]: I would suggest to move individual optimizers into their own files.

File: src/TorchSharp/NN/Optimizer.cs [hunk 5]
```diff
@@ -293,563 +452,1901 @@ public static AdadeltaOptimizer Adadelta(IEnumerable<(string name, Modules.Param
...
+                        param.add_(grad, alpha: alpha);
+
+                    }
+                }, closure);
+            }
+
+            protected override void Dispose(bool disposing)
+            {
+                base.Dispose(disposing);
+                foreach (var (name, state) in _state) {
```
Comment (lostmsu) [inline src/TorchSharp/NN/Optimizer.cs]: NIT: you can do `_state.Values` instead of unpacking `KeyValuePair` here.

Also, where is `_state` defined? When reading this change sequentially, I encountered it a few times by now, but without seeing its definition before the first use I am not certain about its contents.

File: src/TorchSharp/NN/Optimizer.cs [hunk 6]
```diff
@@ -293,563 +452,1901 @@ public static AdadeltaOptimizer Adadelta(IEnumerable<(string name, Modules.Param
...
+                if (!opt.maximize.HasValue) opt.maximize = def.maximize;
+
+                opt.InitialLearningRate = opt.LearningRate.Value;
+
+                _parameter_groups.Add(param_group);
+
+                foreach (var p in param_group.Parameters) {
+                    var state = new State();
+                    _state[p.Handle] = state;
+                    state.momentum_buffer = null;
```
Comment (lostmsu) [inline src/TorchSharp/NN/Optimizer.cs]: NIT: why is this needed?

File: src/TorchSharp/NN/Optimizer.cs [hunk 7]
```diff
@@ -293,563 +452,1901 @@ public static AdadeltaOptimizer Adadelta(IEnumerable<(string name, Modules.Param
...
+                _parameter_groups = new List<Modules.ParamGroup>();
+
+                foreach (var g in parameters) {
+                    add_param_group(g);
+                }
+            }
+
+            public override Tensor step(Func<Tensor> closure = null)
+            {
+                return _step<ParamGroup>(group => {
```
Comment (lostmsu) [inline src/TorchSharp/NN/Optimizer.cs]: I can't really check the correctness of the optimizer implementation.

A few things:

1. It needs a link to either the paper or to the PyTorch implementation
2. It needs a reproducible test for every valid options combination.

**Review by lostmsu — CHANGES_REQUESTED** [review]

File: src/TorchSharp/NN/Module.cs (hunk 1 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Module.cs:157]: Yes, this is the fix to #516

File: src/TorchSharp/NN/LRScheduler.cs (hunk 2 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/LRScheduler.cs:29]: Following the Pytorch API.

https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.StepLR.html#steplr

 The schedulers get the parameter groups from the optimizer.

Comment (lostmsu) [reply src/TorchSharp/NN/LRScheduler.cs:29]: Is there a reason for the scheduler to store a link to the optimizer? If it only needs to store parameter groups internally, this it what is should it. You can have a constructor overload taking optimizer.

File: src/TorchSharp/NN/Optimizer.cs [hunk 8]
```diff
@@ -293,563 +452,1901 @@ public static AdadeltaOptimizer Adadelta(IEnumerable<(string name, Modules.Param
...
+                            : grad.alias();
+
+                        square_avg.mul_(rho).addcmul_(grad, grad, 1 - rho);
+
+                        var std = square_avg.add(eps).sqrt_();
+                        var delta = acc_delta.add(eps).sqrt_().div_(std).mul_(grad);
+
+                        param.add_(delta, alpha: -lr);
+                        acc_delta.mul_(rho).addcmul_(delta, delta, 1 - rho);
+                    }
```
[the remark this answers is not in the document]
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Optimizer.cs]: Yes, that's my conviction. the algorithms of optimizers are very stable.

File: src/TorchSharp/NN/Optimizer.cs (hunk 7 above)
Comment (lostmsu) [reply src/TorchSharp/NN/Optimizer.cs]: Writing tests could be pretty easy with `TestOptimizer(Model, Optimizer, float targetLoss)` helper function that would do a reproducible run.

The problem would be to know what the target loss should be.

File: src/TorchSharp/NN/Optimizer.cs [hunk 9]
```diff
@@ -293,563 +452,1901 @@ public static AdadeltaOptimizer Adadelta(IEnumerable<(string name, Modules.Param
...
+                }, closure);
+            }
+
+            protected override void Dispose(bool disposing)
+            {
+                base.Dispose(disposing);
+                foreach (var (name, state) in _state) {
+                    if (state.momentum_buffer is not null) {
+                        state.momentum_buffer.Dispose();
+                    }
```
[the remark this answers is not in the document]
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Optimizer.cs]: I agree -- that's cleaner. Thx!

File: src/TorchSharp/NN/Optimizer.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (lostmsu) [reply src/TorchSharp/NN/Optimizer.cs]: @NiklasGustafsson frankly, this alone looks like at least 2x in size and probably 10-20x in complexity over mirroring structs in C++.

**Comment by NiklasGustafsson** [comment]
@lostmsu -- your review and comments are much appreciated!

File: test/TorchSharpTest/TestTraining.cs [hunk 10]
```diff
@@ -183,6 +185,43 @@ public void TestTrainingAdamDefaults()
...
+
+                finalLoss = lossVal;
+
+                optimizer.zero_grad();
+
+                output.backward();
+
+                optimizer.step();
+            }
+            Assert.True(finalLoss < initialLoss);
```
[the remark this answers is not in the document]
Comment (NiklasGustafsson) [reply test/TorchSharpTest/TestTraining.cs:222]: I have to think about that. In the past, we've struggled with reproducible RNG sequences in the unit tests, since the tests are run in parallel. As long as there aren't any hidden random number API calls that use the global RNG, it should be doable by using Generator objects.

[the remark this answers is not in the document]
Comment (NiklasGustafsson) [reply test/TorchSharpTest/TestTraining.cs:222]: For example, I don't know how Dropout layers generate the mask, but we can just avoid Dropout in unit tests.

File: src/TorchSharp/NN/LRScheduler.cs (hunk 2 above)
Comment (lostmsu) [reply src/TorchSharp/NN/LRScheduler.cs:29]: Sounds like a pretty good explanation to me.

But if user can remove a parameter group, the corresponding state should be removed as well.

File: src/TorchSharp/NN/Optimizer.cs (hunk 7 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Optimizer.cs]: >> The problem would be to know what the target loss should be.

Yes. I can think of doing it two ways:

1. Write all the unit tests in both Python and .NET and use the Python values as 'expected.' Tons of work.
2. Run the unit tests once, look at the loss and use that as the target loss.

**Comment by GeorgeS2019** [comment]
> @NiklasGustafsson before jumping into this, what is the motivation to move optimizer implementations into C# code? Every implementation copied from/modeled after the original PyTorch code is a maintenance burden, and a huge potential for future divergence from PyTorch in subtle but potentially very important details. Also error prone. That approach and its consequences are one of the major reasons I am not using TensorFlow.NET project.

@lostmsu You made a right observation. One critical difference between TorchSharp and Tensorflow.NET is that the **% of Unit Test code coverage** of LibTorch API and TorchSharp (in my impression) is significantly much higher in TorchSharp compared the situation in Tensorflow.NET.  

The transparency of code coverage and regular alignment with the latest pytorch in Torchsharp is also a key advantage.

File: src/TorchSharp/NN/Optimizer.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (NiklasGustafsson) [reply src/TorchSharp/NN/Optimizer.cs]: Here's a new issue to track: 

https://github.com/dotnet/TorchSharp/issues/531

**Comment by NiklasGustafsson** [comment]
The PR comments, except for the request to revert optimizers to native code, have been taken into account. The optimizer implementation request will be tracked in its own issue (#531) and addressed as long as the long-term cost is not found to be prohibitive.
